tsfc/loopy.py

```diff
@@ -23,6 +22,116 @@
 from contextlib import contextmanager
 
 
+# Table of handled math functions in real and complex modes
+# Note that loopy handles addition of type prefixes and suffixes itself.
+math_table = {
+    'sqrt': ('sqrt', 'sqrt'),
+    'abs': ('abs', 'abs'),
+    'cos': ('cos', 'cos'),
+    'sin': ('sin', 'sin'),
+    'tan': ('tan', 'tan'),
+    'acos': ('acos', 'acos'),
+    'asin': ('asin', 'asin'),
+    'atan': ('atan', 'atan'),
+    'cosh': ('cosh', 'cosh'),
+    'sinh': ('sinh', 'sinh'),
+    'tanh': ('tanh', 'tanh'),
+    'acosh': ('acosh', 'acosh'),
+    'asinh': ('asinh', 'asinh'),
+    'atanh': ('atanh', 'atanh'),
+    'power': ('pow', 'pow'),
+    'exp': ('exp', 'exp'),
+    'ln': ('log', 'log'),
+    'real': (None, 'real'),
+    'imag': (None, 'imag'),
+    'conj': (None, 'conj'),
+    'erf': ('erf', None),
+    'atan_2': ('atan2', None),
+    'atan2': ('atan2', None),
+    'min_value': ('min', None),
+    'max_value': ('max', None)
+}
```

Comment:
This table now feels almost unnecessary.

- The real and complex function names are always the same for those functions which have both versions.
- ... and this target function name is in most cases the same as the source function name, except for `power`/`pow`, `ln`/`log`, `atan_2`/`atan2`, `min_value`/`min`, and `max_value`/`max`.
- Why are there two source names for atan2 by the way?  That looks like an inconsistency somewhere else in the Firedrake stack.
- Power, min value, and max value have their own GEM nodes anyway, so they need not make use of this table.
- That leaves `ln`/`log` alone where a name translation is necessary, plus the dubious atan2 case.

Reply:
I checked `atan_2` is just a mistake. I refactored to remove this table. That also fixes a bug where one couldn't use `erf` (or similar) in complex mode even if the argument was wrapped in `real` to ensure realness.